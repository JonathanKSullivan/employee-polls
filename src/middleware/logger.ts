import { Middleware, AnyAction } from 'redux';

const logger: Middleware = (store) => (next) => (action: unknown) => {
  const typedAction = action as AnyAction;

  console.group(typedAction.type);
  console.log('The action:', typedAction);
  const returnValue = next(typedAction);
  console.log('The new state:', store.getState());
  console.groupEnd();

  return returnValue;
};

export default logger;
