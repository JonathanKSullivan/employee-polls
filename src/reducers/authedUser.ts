interface AuthedUserAction {
  type: 'LOGIN_USER' | 'LOGOUT_USER';
  user?: { id: string };
}

const autheduser = (state: string | null = null, action: AuthedUserAction) => {
  switch (action.type) {
    case 'LOGIN_USER':
      return action.user ? action.user.id : state;
    case 'LOGOUT_USER':
      return null;
    default:
      return state;
  }
};

export default autheduser;
