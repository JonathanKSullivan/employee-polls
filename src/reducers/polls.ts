import { Poll } from '../types/poll';
import { ANSWER_POLL, CREATE_POLL, RECEIVE_POLLS } from '@/actions/polls';

interface PollsAction {
  type: string;
  polls?: { [key: string]: Poll };
  poll?: Poll;
  qid?: string;
  answer?: 'optionOne' | 'optionTwo';
  authedUser?: string;
}

const polls = (state: { [key: string]: Poll } = {}, action: PollsAction) => {
  switch (action.type) {
    case RECEIVE_POLLS:
      return {
        ...state,
        ...action.polls,
      };

    case CREATE_POLL:
      return {
        ...state,
        [action.poll!.id]: action.poll,
      };

    case ANSWER_POLL:
      const { qid, answer, authedUser } = action;
      return {
        ...state,
        [qid!]: {
          ...state[qid!],
          [answer!]: {
            ...state[qid!][answer!],
            votes: state[qid!][answer!].votes.concat([authedUser!]),
          },
        },
      };

    default:
      return state;
  }
};

export default polls;
