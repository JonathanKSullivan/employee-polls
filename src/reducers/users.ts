import { User } from '../types/user';
import { RECEIVE_USERS } from '../actions/users';
import { ANSWER_POLL, CREATE_POLL } from '@/actions/polls';
import { Poll } from '../types/poll';

interface UsersAction {
  type: string;
  users?: { [key: string]: User };
  pollId?: string;
  authedUser?: string;
  answer?: 'optionOne' | 'optionTwo';
  poll?: Poll;
}

const users = (state: { [key: string]: User } = {}, action: UsersAction) => {
  switch (action.type) {
    case RECEIVE_USERS:
      return {
        ...state,
        ...action.users,
      };

    case CREATE_POLL: {
      const { poll } = action;
      return {
        ...state,
        [poll!.author]: {
          ...state[poll!.author],
          questions: state[poll!.author].questions.concat([poll!.id]),
        },
      };
    }

    case ANSWER_POLL: {
      const { pollId, authedUser, answer } = action;
      return {
        ...state,
        [authedUser!]: {
          ...state[authedUser!],
          answers: {
            ...state[authedUser!].answers,
            [pollId!]: answer!,
          },
        },
      };
    }

    default:
      return state;
  }
};

export default users;
