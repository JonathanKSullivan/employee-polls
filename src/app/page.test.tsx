import React from 'react';
import { Provider } from 'react-redux';
import { render, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import { createStore } from 'redux';
import reducers from '@/reducers';
import middleware from '@/middleware';
import Home from './page';
import { useRouter } from 'next/navigation';
import renderer from 'react-test-renderer';

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));

const store = createStore(
  reducers,
  { users: {}, polls: {}, authedUser: null },
  middleware
);

describe('Home Page', () => {
  beforeEach(() => {
    const mockPush = jest.fn();
    (useRouter as jest.Mock).mockReturnValue({
      push: mockPush,
      pathname: '/',
    });
  });

  it('should render the home page', () => {
    const { getByText } = render(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    expect(getByText('Welcome to the Polling App!')).toBeInTheDocument();
  });

  it('should match the snapshot', () => {
    let component;

    act(() => {
      component = renderer.create(
        <Provider store={store}>
          <Home />
        </Provider>
      );
    });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
