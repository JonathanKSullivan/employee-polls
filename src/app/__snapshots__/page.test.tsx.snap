// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page should match the snapshot 1`] = `
<div>
  <div
    className="ant-layout css-dev-only-do-not-override-ccdg5a"
    style={{}}
  >
    <header
      className="ant-layout-header css-dev-only-do-not-override-ccdg5a"
    >
      <ul
        className="ant-menu-overflow ant-menu ant-menu-root ant-menu-horizontal ant-menu-dark css-dev-only-do-not-override-ccdg5a"
        data-menu-list={true}
        onKeyDown={[Function]}
        role="menu"
        style={{}}
        tabIndex={0}
      >
        <li
          className="ant-menu-overflow-item ant-menu-item ant-menu-item-selected"
          data-menu-id="rc-menu-uuid-test-home"
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onPointerEnter={[Function]}
          onPointerLeave={[Function]}
          role="menuitem"
          style={
            {
              "height": undefined,
              "opacity": 1,
              "order": 0,
              "overflowY": undefined,
              "pointerEvents": undefined,
              "position": undefined,
            }
          }
          tabIndex={-1}
        >
          <span
            aria-label="home"
            className="anticon anticon-home ant-menu-item-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="home"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M946.5 505L560.1 118.8l-25.9-25.9a31.5 31.5 0 00-44.4 0L77.5 505a63.9 63.9 0 00-18.8 46c.4 35.2 29.7 63.3 64.9 63.3h42.5V940h691.8V614.3h43.4c17.1 0 33.2-6.7 45.3-18.8a63.6 63.6 0 0018.7-45.3c0-17-6.7-33.1-18.8-45.2zM568 868H456V664h112v204zm217.9-325.7V868H632V640c0-22.1-17.9-40-40-40H432c-22.1 0-40 17.9-40 40v228H238.1V542.3h-96l370-369.7 23.1 23.1L882 542.3h-96.1z"
              />
            </svg>
          </span>
          <span
            className="ant-menu-title-content"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              Home
            </a>
          </span>
        </li>
        <li
          className="ant-menu-overflow-item ant-menu-item"
          data-menu-id={null}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onPointerEnter={[Function]}
          onPointerLeave={[Function]}
          role="menuitem"
          style={
            {
              "height": undefined,
              "opacity": 1,
              "order": 1,
              "overflowY": undefined,
              "pointerEvents": undefined,
              "position": undefined,
            }
          }
          tabIndex={-1}
        >
          <span
            aria-label="plus"
            className="anticon anticon-plus ant-menu-item-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="plus"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
              />
              <path
                d="M192 474h672q8 0 8 8v60q0 8-8 8H160q-8 0-8-8v-60q0-8 8-8z"
              />
            </svg>
          </span>
          <span
            className="ant-menu-title-content"
          >
            <a
              href="/add"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              Add
            </a>
          </span>
        </li>
        <li
          className="ant-menu-overflow-item ant-menu-item"
          data-menu-id={null}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onPointerEnter={[Function]}
          onPointerLeave={[Function]}
          role="menuitem"
          style={
            {
              "height": undefined,
              "opacity": 1,
              "order": 2,
              "overflowY": undefined,
              "pointerEvents": undefined,
              "position": undefined,
            }
          }
          tabIndex={-1}
        >
          <span
            aria-label="trophy"
            className="anticon anticon-trophy ant-menu-item-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="trophy"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M868 160h-92v-40c0-4.4-3.6-8-8-8H256c-4.4 0-8 3.6-8 8v40h-92a44 44 0 00-44 44v148c0 81.7 60 149.6 138.2 162C265.7 630.2 359 721.7 476 734.5v105.2H280c-17.7 0-32 14.3-32 32V904c0 4.4 3.6 8 8 8h512c4.4 0 8-3.6 8-8v-32.3c0-17.7-14.3-32-32-32H548V734.5C665 721.7 758.3 630.2 773.8 514 852 501.6 912 433.7 912 352V204a44 44 0 00-44-44zM184 352V232h64v207.6a91.99 91.99 0 01-64-87.6zm520 128c0 49.1-19.1 95.4-53.9 130.1-34.8 34.8-81 53.9-130.1 53.9h-16c-49.1 0-95.4-19.1-130.1-53.9-34.8-34.8-53.9-81-53.9-130.1V184h384v296zm136-128c0 41-26.9 75.8-64 87.6V232h64v120z"
              />
            </svg>
          </span>
          <span
            className="ant-menu-title-content"
          >
            <a
              href="/leaderboard"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              Leaderboard
            </a>
          </span>
        </li>
        <li
          className="ant-menu-overflow-item ant-menu-item"
          data-menu-id={null}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onPointerEnter={[Function]}
          onPointerLeave={[Function]}
          role="menuitem"
          style={
            {
              "height": undefined,
              "marginLeft": "auto",
              "opacity": 1,
              "order": 3,
              "overflowY": undefined,
              "pointerEvents": undefined,
              "position": undefined,
            }
          }
          tabIndex={-1}
        >
          <span
            aria-label="user"
            className="anticon anticon-user ant-menu-item-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="user"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
              />
            </svg>
          </span>
          <span
            className="ant-menu-title-content"
          >
            <a
              href="/login"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              <button
                className="ant-btn css-dev-only-do-not-override-ccdg5a ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                disabled={false}
                onClick={[Function]}
                style={{}}
                type="button"
              >
                <span>
                  Login
                </span>
              </button>
            </a>
          </span>
        </li>
        <li
          aria-hidden={true}
          className="ant-menu-overflow-item ant-menu-overflow-item-rest ant-menu-submenu ant-menu-submenu-horizontal"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          role="none"
          style={
            {
              "height": 0,
              "opacity": 0,
              "order": 9007199254740991,
              "overflowY": "hidden",
              "pointerEvents": "none",
              "position": "absolute",
            }
          }
        >
          <div
            aria-controls="rc-menu-uuid-test-rc-menu-more-popup"
            aria-expanded={false}
            aria-haspopup={true}
            className="ant-menu-submenu-title"
            data-menu-id="rc-menu-uuid-test-rc-menu-more"
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onPointerEnter={[Function]}
            onPointerLeave={[Function]}
            role="menuitem"
            style={null}
            tabIndex={-1}
            title={null}
          >
            <span
              aria-label="ellipsis"
              className="anticon anticon-ellipsis"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="ellipsis"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
                />
              </svg>
            </span>
            <i
              className="ant-menu-submenu-arrow"
            />
          </div>
        </li>
      </ul>
      <div
        aria-hidden={true}
        style={
          {
            "display": "none",
          }
        }
      />
    </header>
  </div>
  <div
    style={
      {
        "marginTop": "50px",
        "textAlign": "center",
      }
    }
  >
    <h1
      className="ant-typography css-dev-only-do-not-override-ccdg5a"
      style={
        {
          "WebkitLineClamp": undefined,
        }
      }
    >
      Welcome to the Polling App!
    </h1>
    <div
      className="ant-typography css-dev-only-do-not-override-ccdg5a"
      style={
        {
          "WebkitLineClamp": undefined,
        }
      }
    >
      This is a community-driven app where you can participate in fun and thought-provoking polls. Answer existing polls, create your own, and see how others voted.
    </div>
    <div
      className="ant-typography css-dev-only-do-not-override-ccdg5a"
      style={
        {
          "WebkitLineClamp": undefined,
        }
      }
    >
      To get started, please 
      <span
        className="ant-typography css-dev-only-do-not-override-ccdg5a"
        style={
          {
            "WebkitLineClamp": undefined,
          }
        }
      >
        <strong>
          log in
        </strong>
      </span>
       or
       
      <span
        className="ant-typography css-dev-only-do-not-override-ccdg5a"
        style={
          {
            "WebkitLineClamp": undefined,
          }
        }
      >
        <strong>
          sign up
        </strong>
      </span>
       and join the conversation!
    </div>
    <button
      className="ant-btn css-dev-only-do-not-override-ccdg5a ant-btn-primary ant-btn-color-primary ant-btn-variant-solid ant-btn-lg"
      disabled={false}
      onClick={[Function]}
      style={
        {
          "marginTop": "20px",
        }
      }
      type="button"
    >
      <span>
        Log In
      </span>
    </button>
  </div>
</div>
`;
